staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go

```diff
@@ -383,51 +383,98 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
 		if pool.Generation > generation {
 			generation = pool.Generation
 		}
-
-		// Right now all devices get published in a single slice.
-		// We simply pick the first one, if there is one, and copy
-		// it in preparation for updating it.
-		//
-		// TODO: support splitting across slices, with unit tests.
-		if len(slices) > 0 {
-			obsoleteSlices = append(obsoleteSlices, slices[1:]...)
-			slices = []*resourceapi.ResourceSlice{slices[0].DeepCopy()}
-		} else {
-			slices = []*resourceapi.ResourceSlice{
-				{
+		needToAdd, needToDelete := getDevicesToAddAndDelete(pool, slices)
+		c.mutex.Lock()
+		slices = removeDevicesFromSlice(slices, needToDelete)
+		c.mutex.Unlock()
+		// Add new devices to slices.
+		for _, device := range needToAdd {
+			added := false
+			for i := 0; i < len(slices); i++ {
+				// If a slice is not full, it can be added to.
+				if len(slices[i].Spec.Devices) < resourceapi.ResourceSliceMaxDevices {
+					slices[i].Spec.Devices = append(slices[i].Spec.Devices, device)
```

Comment:
This is a more likely reason for concurrency problems: the *content* of slices is from `c.sliceStore.ByIndex(poolNameIndex, poolName)`. In other words: the shared cache. Such objects must not be modified.

Before changing them, you must do a `DeepCopy`.

Reply:
sorry for this. I will fixed this